<a href="https://colab.research.google.com/github/rahajengs/UTS-PMML-Rahajeng-Septri-Annisa/blob/main/UTS_PMML_Rahajeng_Septri_Annisa.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Nama: Rahajeng Septri Annisa


NIM: 2404220036


Prodi: Statistika dan Sains Data


UTS Prediksi Modern dan Machine Learning

#  Data Understanding & Preparation

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split, StratifiedKFold, GridSearchCV, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report, confusion_matrix

In [2]:
from google.colab import drive

drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/UTS PMML/UCI_Credit_Card.csv')
data

,ID,LIMIT_BAL,SEX,EDUCATION,MARRIAGE,AGE,PAY_0,PAY_2,PAY_3,PAY_4,...,BILL_AMT4,BILL_AMT5,BILL_AMT6,PAY_AMT1,PAY_AMT2,PAY_AMT3,PAY_AMT4,PAY_AMT5,PAY_AMT6,default.payment.next.month
0,1,20000.0,2,2,1,24,2,2,-1,-1,...,0.0,0.0,0.0,0.0,689.0,0.0,0.0,0.0,0.0,1
1,2,120000.0,2,2,2,26,-1,2,0,0,...,3272.0,3455.0,3261.0,0.0,1000.0,1000.0,1000.0,0.0,2000.0,1
2,3,90000.0,2,2,2,34,0,0,0,0,...,14331.0,14948.0,15549.0,1518.0,1500.0,1000.0,1000.0,1000.0,5000.0,0
3,4,50000.0,2,2,1,37,0,0,0,0,...,28314.0,28959.0,29547.0,2000.0,2019.0,1200.0,1100.0,1069.0,1000.0,0
4,5,50000.0,1,2,1,57,-1,0,-1,0,...,20940.0,19146.0,19131.0,2000.0,36681.0,10000.0,9000.0,689.0,679.0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
29995,29996,220000.0,1,3,1,39,0,0,0,0,...,88004.0,31237.0,15980.0,8500.0,20000.0,5003.0,3047.0,5000.0,1000.0,0
29996,29997,150000.0,1,3,2,43,-1,-1,-1,-1,...,8979.0,5190.0,0.0,1837.0,3526.0,8998.0,129.0,0.0,0.0,0
29997,29998,30000.0,1,2,2,37,4,3,2,-1,...,20878.0,20582.0,19357.0,0.0,0.0,22000.0,4200.0,2000.0,3100.0,1
29998,29999,80000.0,1,3,1,41,1,-1,0,0,...,52774.0,11855.0,48944.0,85900.0,3409.0,1178.0,1926.0,52964.0,1804.0,1


In [4]:
#melihat struktur dan tipe data
data.info()
data.dtypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   ID                          30000 non-null  int64  
 1   LIMIT_BAL                   30000 non-null  float64
 2   SEX                         30000 non-null  int64  
 3   EDUCATION                   30000 non-null  int64  
 4   MARRIAGE                    30000 non-null  int64  
 5   AGE                         30000 non-null  int64  
 6   PAY_0                       30000 non-null  int64  
 7   PAY_2                       30000 non-null  int64  
 8   PAY_3                       30000 non-null  int64  
 9   PAY_4                       30000 non-null  int64  
 10  PAY_5                       30000 non-null  int64  
 11  PAY_6                       30000 non-null  int64  
 12  BILL_AMT1                   30000 non-null  float64
 13  BILL_AMT2                   300

,0
ID,int64
LIMIT_BAL,float64
SEX,int64
EDUCATION,int64
MARRIAGE,int64
AGE,int64
PAY_0,int64
PAY_2,int64
PAY_3,int64
PAY_4,int64


In [5]:
target = 'default.payment.next.month'
print(data[target].value_counts())

default.payment.next.month
0    23364
1     6636
Name: count, dtype: int64


Target:

0 = tidak mengalami default
1 = mengalami default

In [6]:
#periksa missing value
data.isnull().sum()

,0
ID,0
LIMIT_BAL,0
SEX,0
EDUCATION,0
MARRIAGE,0
AGE,0
PAY_0,0
PAY_2,0
PAY_3,0
PAY_4,0


In [7]:
#periksa duplicate
data.duplicated().sum()

np.int64(0)

In [8]:
#Nilai unik beberapa variabel kategorikal/kode
categorical_code = [
    'SEX',
    'EDUCATION',
    'MARRIAGE',
    'PAY_0',
    'PAY_2',
    'PAY_3',
    'PAY_4',
    'PAY_5',
    'PAY_6'
]

for col in categorical_code:
    print(f"\n{col}:")
    print(sorted(data[col].unique()))

print("EDUCATION:")
print(data['EDUCATION'].value_counts().sort_index())

print("\nMARRIAGE:")
print(data['MARRIAGE'].value_counts().sort_index())


SEX:
[np.int64(1), np.int64(2)]

EDUCATION:
[np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6)]

MARRIAGE:
[np.int64(0), np.int64(1), np.int64(2), np.int64(3)]

PAY_0:
[np.int64(-2), np.int64(-1), np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8)]

PAY_2:
[np.int64(-2), np.int64(-1), np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8)]

PAY_3:
[np.int64(-2), np.int64(-1), np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8)]

PAY_4:
[np.int64(-2), np.int64(-1), np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8)]

PAY_5:
[np.int64(-2), np.int64(-1), np.int64(0), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8)]

PAY_6:
[np.int64(-2), np.int64(-1), np.in

In [9]:
#menentukan target
y = data['default.payment.next.month']

#menentukan predictor
X = data.drop(columns=['default.payment.next.month'])

In [10]:
#menghapus ID dari predictor
X = X.drop(columns=['ID'])

In [11]:
#split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

In [12]:
logistic_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('model', LogisticRegression(
        max_iter=2000,
        solver='liblinear'))
])

# BASELINE (LOGISTIC REGRESSION)

In [13]:
#membuat model Logistic Regression
baseline_model = LogisticRegression(
    max_iter=2000,
    solver='liblinear',
    random_state=42
)

#5-Fold Cross Validation
cv = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

#menghitung CV F1-score
cv_f1_scores = cross_val_score(
    baseline_model,
    X_train,
    y_train,
    cv=cv,
    scoring='f1'
)

print("F1-score setiap fold:")
for i, score in enumerate(cv_f1_scores, start=1):
    print(f"Fold {i}: {score:.4f}")

print(f"\nMean CV F1-score: {cv_f1_scores.mean():.4f}")
print(f"Std CV F1-score : {cv_f1_scores.std():.4f}")

F1-score setiap fold:
Fold 1: 0.0000
Fold 2: 0.0000
Fold 3: 0.0000
Fold 4: 0.0000
Fold 5: 0.0000

Mean CV F1-score: 0.0000
Std CV F1-score : 0.0000


In [14]:
#training model pada training set
baseline_model.fit(X_train, y_train)

LogisticRegression(max_iter=2000, random_state=42, solver='liblinear')

In [15]:
#prediksi pada data test
y_pred_baseline = baseline_model.predict(X_test)

In [16]:
#menghitung metrik evaluasi test set
accuracy_baseline = accuracy_score(y_test, y_pred_baseline)

precision_baseline = precision_score(
    y_test,
    y_pred_baseline,
    zero_division=0
)

recall_baseline = recall_score(
    y_test,
    y_pred_baseline,
    zero_division=0
)

f1_baseline = f1_score(
    y_test,
    y_pred_baseline,
    zero_division=0
)

baseline_results = pd.DataFrame({
    'Model': ['Baseline Logistic Regression'],
    'CV F1-score': [cv_f1_scores.mean()],
    'Test Accuracy': [accuracy_baseline],
    'Test Precision': [precision_baseline],
    'Test Recall': [recall_baseline],
    'Test F1-score': [f1_baseline]
})

baseline_results

,Model,CV F1-score,Test Accuracy,Test Precision,Test Recall,Test F1-score
0,Baseline Logistic Regression,0.0,0.778833,0.0,0.0,0.0


Berdasarkan hasil Experiment 1, Logistic Regression sebagai baseline menghasilkan nilai Test Accuracy sebesar 0,7788, sedangkan Precision, Recall, dan F1-score masing-masing sebesar 0,0. Nilai CV F1-score juga sebesar 0,0. Hasil tersebut menunjukkan bahwa model baseline belum mampu mengidentifikasi kelas default dengan baik. Logistic Regression tanpa preprocessing digunakan sebagai baseline untuk membandingkan performa pada eksperimen berikutnya.

# Preprocessing (Logistic Regression + Standard Scaler)

In [17]:
#membuat pipeline preprocessing + model
logreg_scaled = Pipeline([
    ('scaler', StandardScaler()),
    ('logistic', LogisticRegression(
        max_iter=2000,
        solver='liblinear',
        random_state=42
    ))
])

logreg_scaled

Pipeline(steps=[('scaler', StandardScaler()),
                ('logistic',
                 LogisticRegression(max_iter=2000, random_state=42,
                                    solver='liblinear'))])

In [18]:
cv_f1_exp2 = cross_val_score(
    logreg_scaled,
    X_train,
    y_train,
    cv=cv,
    scoring='f1'
)

print("F1-score setiap fold:")
for i, score in enumerate(cv_f1_exp2, 1):
    print(f"Fold {i}: {score:.4f}")

print(f"\nMean CV F1-score: {cv_f1_exp2.mean():.4f}")

F1-score setiap fold:
Fold 1: 0.3669
Fold 2: 0.3732
Fold 3: 0.3804
Fold 4: 0.3638
Fold 5: 0.3407

Mean CV F1-score: 0.3650


In [19]:
#training model
logreg_scaled.fit(X_train, y_train)

Pipeline(steps=[('scaler', StandardScaler()),
                ('logistic',
                 LogisticRegression(max_iter=2000, random_state=42,
                                    solver='liblinear'))])

In [20]:
#prediksi data test
y_pred_exp2 = logreg_scaled.predict(X_test)

In [21]:
#menghitung metrik
accuracy_exp2 = accuracy_score(y_test, y_pred_exp2)

precision_exp2 = precision_score(
    y_test,
    y_pred_exp2,
    zero_division=0
)

recall_exp2 = recall_score(
    y_test,
    y_pred_exp2,
    zero_division=0
)

f1_exp2 = f1_score(
    y_test,
    y_pred_exp2,
    zero_division=0
)

exp2_results = pd.DataFrame({
    'Model': ['Logistic Regression + Standard Scaler'],
    'CV F1-score': [cv_f1_exp2.mean()],
    'Test Accuracy': [accuracy_exp2],
    'Test Precision': [precision_exp2],
    'Test Recall': [recall_exp2],
    'Test F1-score': [f1_exp2]
})

exp2_results

,Model,CV F1-score,Test Accuracy,Test Precision,Test Recall,Test F1-score
0,Logistic Regression + Standard Scaler,0.365009,0.807833,0.688312,0.239638,0.355506


In [22]:
# Membuat tabel perbandingan
comparison_exp1_exp2 = pd.DataFrame({
    'Experiment': [
        'Experiment 1',
        'Experiment 2'
    ],
    'Model': [
        'Logistic Regression',
        'Logistic Regression + StandardScaler'
    ],
    'CV F1-score': [
        cv_f1_scores.mean(),
        cv_f1_exp2.mean()
    ],
    'Test Accuracy': [
        accuracy_baseline,
        accuracy_exp2
    ],
    'Test Precision': [
        precision_baseline,
        precision_exp2
    ],
    'Test Recall': [
        recall_baseline,
        recall_exp2
    ],
    'Test F1-score': [
        f1_baseline,
        f1_exp2
    ]
})

comparison_exp1_exp2

,Experiment,Model,CV F1-score,Test Accuracy,Test Precision,Test Recall,Test F1-score
0,Experiment 1,Logistic Regression,0.000000,0.778833,0.000000,0.000000,0.000000
1,Experiment 2,Logistic Regression + StandardScaler,0.365009,0.807833,0.688312,0.239638,0.355506


Berdasarkan hasil eksperimen, preprocessing menggunakan StandardScaler memberikan peningkatan performa Logistic Regression dibandingkan baseline.

# Model Comparison

In [23]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

MODEL 2 (DECISION TREE)

In [24]:
decision_tree = DecisionTreeClassifier(random_state=42)

In [25]:
cv_f1_tree = cross_val_score(
    decision_tree,
    X_train,
    y_train,
    cv=cv,
    scoring='f1'
)

print("Decision Tree")
print("CV F1-score setiap fold:")

for i, score in enumerate(cv_f1_tree, 1):
    print(f"Fold {i}: {score:.4f}")

print(f"Mean CV F1-score: {cv_f1_tree.mean():.4f}")

Decision Tree
CV F1-score setiap fold:
Fold 1: 0.4176
Fold 2: 0.4045
Fold 3: 0.4055
Fold 4: 0.3935
Fold 5: 0.3890
Mean CV F1-score: 0.4020


In [26]:
#Training
decision_tree.fit(X_train, y_train)

#Prediksi
y_pred_tree = decision_tree.predict(X_test)

#Evaluasi
acc_tree = accuracy_score(y_test, y_pred_tree)
prec_tree = precision_score(y_test, y_pred_tree, zero_division=0)
rec_tree = recall_score(y_test, y_pred_tree, zero_division=0)
f1_tree = f1_score(y_test, y_pred_tree, zero_division=0)

print("\n=== Decision Tree ===")
print(f"Accuracy    : {acc_tree:.4f}")
print(f"Precision   : {prec_tree:.4f}")
print(f"Recall      : {rec_tree:.4f}")
print(f"F1-score    : {f1_tree:.4f}")


=== Decision Tree ===
Accuracy    : 0.7145
Precision   : 0.3694
Recall      : 0.4115
F1-score    : 0.3893


MODEL 3 (RANDOM FOREST)

In [27]:
random_forest = RandomForestClassifier(random_state=42, n_jobs=-1)

In [28]:
cv_f1_rf = cross_val_score(
    random_forest,
    X_train,
    y_train,
    cv=cv,
    scoring='f1'
)

print("Random Forest")
print("CV F1-score setiap fold:")

for i, score in enumerate(cv_f1_rf, 1):
    print(f"Fold {i}: {score:.4f}")

print(f"Mean CV F1-score: {cv_f1_rf.mean():.4f}")

Random Forest
CV F1-score setiap fold:
Fold 1: 0.4858
Fold 2: 0.4979
Fold 3: 0.4730
Fold 4: 0.4823
Fold 5: 0.4648
Mean CV F1-score: 0.4808


In [29]:
#Training
random_forest.fit(X_train, y_train)

#Prediksi
y_pred_rf = random_forest.predict(X_test)

#Evaluasi
acc_rf = accuracy_score(y_test, y_pred_rf)
prec_rf = precision_score(y_test, y_pred_rf, zero_division=0)
rec_rf = recall_score(y_test, y_pred_rf, zero_division=0)
f1_rf = f1_score(y_test, y_pred_rf, zero_division=0)

print("\n=== Random Forest ===")
print(f"Accuracy    : {acc_rf:.4f}")
print(f"Precision   : {prec_rf:.4f}")
print(f"Recall      : {rec_rf:.4f}")
print(f"F1-score    : {f1_rf:.4f}")


=== Random Forest ===
Accuracy    : 0.8120
Precision   : 0.6325
Recall      : 0.3580
F1-score    : 0.4572


Perbandingan 3 Model

In [30]:
# Membuat tabel hasil perbandingan
hasil_exp3 = pd.DataFrame({
    'Model': [
        'Logistic Regression',
        'Decision Tree',
        'Random Forest'
    ],
    'CV F1-score': [
        cv_f1_exp2.mean(),
        cv_f1_tree.mean(),
        cv_f1_rf.mean()
    ],
    'Test Accuracy': [
        accuracy_exp2,
        acc_tree,
        acc_rf
    ],
    'Test Precision': [
        precision_exp2,
        prec_tree,
        prec_rf
    ],
    'Test Recall': [
        recall_exp2,
        rec_tree,
        rec_rf
    ],
    'Test F1-score': [
        f1_exp2,
        f1_tree,
        f1_rf
    ]
})

hasil_exp3

,Model,CV F1-score,Test Accuracy,Test Precision,Test Recall,Test F1-score
0,Logistic Regression,0.365009,0.807833,0.688312,0.239638,0.355506
1,Decision Tree,0.402027,0.714500,0.369418,0.411454,0.389305
2,Random Forest,0.480764,0.812000,0.632490,0.357950,0.457170


In [31]:
#Menentukan model terbaik
best_model_exp3 = hasil_exp3.loc[
    hasil_exp3['Test F1-score'].idxmax()
]

print("Model terbaik Experiment 3:")
print(best_model_exp3)

Model terbaik Experiment 3:
Model             Random Forest
CV F1-score            0.480764
Test Accuracy             0.812
Test Precision          0.63249
Test Recall             0.35795
Test F1-score           0.45717
Name: 2, dtype: object


Berdasarkan hasil Experiment 3, dilakukan perbandingan antara Logistic Regression, Decision Tree, dan Random Forest menggunakan 5-fold Cross Validation serta evaluasi pada test set. Berdasarkan nilai CV F1-score dan Test F1-score, model Random Forest memberikan performa terbaik dengan CV F1-score sebesar  0.480764 dan Test F1-score sebesar 0.45717. Oleh karena itu, model Random Forest dipilih sebagai model terbaik pada Experiment 3 dan akan digunakan pada Experiment 4 untuk dilakukan hyperparameter tuning.

# Hyperparameter Tuning.

In [32]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(
    random_state=42,
    n_jobs=-1
)

In [33]:
param_grid_rf = {
    'n_estimators': [100, 200],
    'max_depth': [10, 20]
}

In [ ]:
grid_search = GridSearchCV(
    estimator=rf,
    param_grid=param_grid_rf,
    scoring='f1',
    cv=5,
    n_jobs=-1,
    verbose=1
)

# Training GridSearchCV
grid_search.fit(X_train, y_train)

Fitting 5 folds for each of 4 candidates, totalling 20 fits


In [ ]:
print("Hyperparameter terbaik:")
print(grid_search.best_params_)

In [ ]:
print("Best CV F1-score:")
print(f"{grid_search.best_score_:.4f}")

In [ ]:
best_rf = grid_search.best_estimator_

In [ ]:
#Prediksi data test
y_pred_tuned = best_rf.predict(X_test)

In [ ]:
accuracy_tuned = accuracy_score(
    y_test,
    y_pred_tuned
)

precision_tuned = precision_score(
    y_test,
    y_pred_tuned,
    zero_division=0
)

recall_tuned = recall_score(
    y_test,
    y_pred_tuned,
    zero_division=0
)

f1_tuned = f1_score(
    y_test,
    y_pred_tuned,
    zero_division=0
)

print("=== HASIL EXPERIMENT 4 ===")
print(f"Best CV F1-score : {grid_search.best_score_:.4f}")
print(f"Test Accuracy    : {accuracy_tuned:.4f}")
print(f"Test Precision   : {precision_tuned:.4f}")
print(f"Test Recall      : {recall_tuned:.4f}")
print(f"Test F1-score    : {f1_tuned:.4f}")

Membandingkan Sebelum dan Sesudah Tuning

In [ ]:
comparison_tuning = pd.DataFrame({
    'Model': [
        'Random Forest sebelum tuning',
        'Random Forest setelah tuning'
    ],
    'CV F1-score': [
        cv_f1_rf.mean(),
        grid_search.best_score_
    ],
    'Test Accuracy': [
        acc_rf,
        accuracy_tuned
    ],
    'Test Precision': [
        prec_rf,
        precision_tuned
    ],
    'Test Recall': [
        rec_rf,
        recall_tuned
    ],
    'Test F1-score': [
        f1_rf,
        f1_tuned
    ]
})

comparison_tuning

In [ ]:
f1_change = f1_tuned - f1_rf

print(f"Perubahan Test F1-score: {f1_change:+.4f}")

Hyperparameter tuning memberikan peningkatan kecil pada performa test set. Test F1-score meningkat dari 0,4572 menjadi 0,4603, disertai peningkatan Test Accuracy dan Test Precision. Namun, CV F1-score menurun dari 0,4808 menjadi 0,4691 dan Test Recall sedikit menurun. Oleh karena itu, tuning memberikan peningkatan pada beberapa metrik test set, tetapi peningkatannya belum konsisten pada seluruh metrik.

# Advanced Model (XGBoost)

In [ ]:
import xgboost as xgb

In [ ]:
xgb_model = xgb.XGBClassifier(
    n_estimators=200,
    learning_rate=0.05,
    max_depth=5,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=42,
    eval_metric='logloss'
)

In [ ]:
cv_f1_xgb = cross_val_score(
    xgb_model,
    X_train,
    y_train,
    cv=cv,
    scoring='f1',
    n_jobs=-1
)

print("F1-score setiap fold:")
print(cv_f1_xgb)

print("\nRata-rata CV F1-score:")
print(cv_f1_xgb.mean())

In [ ]:
#training XGBoost
xgb_model.fit(X_train, y_train)

In [ ]:
#prediksi data testing
y_pred_xgb = xgb_model.predict(X_test)

In [ ]:
accuracy_xgb = accuracy_score(y_test, y_pred_xgb)
precision_xgb = precision_score(y_test, y_pred_xgb, zero_division=0)
recall_xgb = recall_score(y_test, y_pred_xgb, zero_division=0)
f1_xgb = f1_score(y_test, y_pred_xgb, zero_division=0)

hasil_exp5 = pd.DataFrame({
    'Model': ['XGBoost'],
    'CV F1-score': [cv_f1_xgb.mean()],
    'Test Accuracy': [accuracy_xgb],
    'Test Precision': [precision_xgb],
    'Test Recall': [recall_xgb],
    'Test F1-score': [f1_xgb]
})

hasil_exp5

In [ ]:
#perbandingan Model Experiment 4 (Random Forest setelah tuning) dan XGBoost
perbandingan_exp4_exp5 = pd.concat(
    [
        comparison_tuning.tail(1),
        hasil_exp5
    ],
    ignore_index=True
)

perbandingan_exp4_exp5

Berdasarkan hasil Experiment 5, XGBoost menghasilkan CV F1-score sebesar 0,4781 dan Test F1-score sebesar 0,4709. Nilai tersebut lebih tinggi dibandingkan Random Forest setelah tuning yang memiliki CV F1-score 0,4691 dan Test F1-score 0,4603. Selain itu, XGBoost juga menghasilkan Test Accuracy, Precision, dan Recall yang lebih tinggi. Dengan demikian, berdasarkan hasil eksperimen dan F1-score sebagai metrik utama, XGBoost dipilih sebagai model terbaik.